src/prefect/utilities/asyncutils.py

```diff
@@ -181,8 +182,20 @@ def sync_compatible(async_fn: T) -> T:
     @wraps(async_fn)
     def wrapper(*args, **kwargs):
         if in_async_main_thread():
-            # In the main async context; return the coro for them to await
-            return async_fn(*args, **kwargs)
+            caller_frame = sys._getframe(1)
+            caller_async = caller_frame.f_code.co_flags & inspect.CO_COROUTINE
+            if caller_async:
+                # In the main async context; return the coro for them to await
+                return async_fn(*args, **kwargs)
+            else:
+                # In the main thread but call was made from a sync method
+                raise RuntimeError(
+                    "A 'sync_compatible' method was called from a context that was "
+                    "previously async but is now sync. The sync call must be changed "
+                    "to run in a worker thread to support sending the coroutine for "
```

Comment:
is there any way to simplify this message in an actionable way?  I believe this message is _most_ helpful for folks who are calling this and don't fully understand sync / async

Reply:
Yeah I'm not excited about this message either. I think that it should mostly be internal facing — if a user encounters this, we've failed somehow. Even with this message, it seems like an improvement over just receiving a coroutine. Happy for any better message suggestions though.

Reply:
Maybe write up some documentation on this aspect and have this message point at that documentation?

Reply:
That makes sense to me. I'd prefer to do so in a follow-up though as it'll be a significant amount of work to explain how we're doing sync/async compatibility